Add AWS Fraud Detector

donglaiduan opened the pull request.

donglaiduan committed:
test add aws fraud detector

donglaiduan committed:
change csv to match fraud detector

jlarsen0206 commented on internal/services/aws_fraud_service.go:
```diff
@@ -0,0 +1,145 @@
+			}
+
+			// Call the API
+			result, err := fs.Client.GetEventPrediction(ctx, input)
```
Could you move some of this to another file? Something like internal/fraud_detection/fraud_detector.go

jlarsen0206 commented on internal/services/aws_fraud_service.go:
```diff
@@ -0,0 +1,145 @@
+					"customer_occupation":  txn.CustomerOccupation,
+					"login_attempts":       fmt.Sprintf("%d", txn.LoginAttempts),
+				},
+			}
```
I recommend moving the fraud detector prediction into a GfFraudPredictor.PredictFraud method into a new file, with a FraudDetector interface and GfFraudDetector struct (example below). This is similar to the pattern we follow in many other files other files (repository.go, fraud_service.go, etc.). The benefit of this is that we can pass a mock FraudDetector object for the unit tests and isolate our business logic from the AWS Fraud Detector result. We don't want our unit tests to actually hit AWS when this fraud_service is tested (potentially expensive) and it gives an unpredictable result (it's hard to assert an outcome in tests if we don't know what the model will predict). 

You can do something like this:
internal/fraud_detection/fraud_detector.go
```go
// interface to pass to GfFraudService
type FraudDetector interface {
  PredictFraud(txn models.Transaction) (bool, error)
}

// struct implementing the interface
type GfAWSFraudDetector struct {
  client *frauddetector.Client
}

func NewGfAWSFraudDetector(client *frauddetector.Client) *GfAWSFraudDetector {...}

// Your AWS Fraud Detector code goes here
func(fd *GfAWSFraudDetector) PredictFraud(txn models.Transaction) (bool, error) {...}
```

donglaiduan committed:
Add AWS Fraud Detection Integration

- Introduced a new fraud detection module with GfAWSFraudDetector implementation.
- Updated fraud_pipeline.go to initialize the fraud detector client and integrate it into the AWSFraudService.
- Refactored PredictFraud method to utilize the new fraud detection logic.
- Enhanced error handling and logging for fraud prediction failures.

donglaiduan committed:
lint

donglaiduan committed:
lint

jlarsen0206 commented on cmd/lambda/fraud/fraud_pipeline.go:
```diff
@@ -56,10 +65,13 @@ func main() {
 	eventDispatcher := events.NewGfEventDispatcher(snsMessenger)
-	fraudService := services.NewFraudService(eventDispatcher, repository)
+	fraudDetector := fraud_detection.NewGfAWSFraudDetector(fraudDetectorClient)
+	fraudService := services.NewAWSFraudService(eventDispatcher, repository, fraudDetector)
```
We have unit tests written for the existing FraudService already, and AWSFraudService uses untested business logic. I'd recommend moving your call to PredictFraud into the existing FraudService so that we are using tested code in our deployment

jlarsen0206 commented on internal/services/aws_fraud_service.go:
```diff
@@ -0,0 +1,89 @@
+		approvedTransactions = append(approvedTransactions, txn)
+	}
+
+	return fraudulentTransactions, approvedTransactions, middleware.MergeErrors(errorResults)
```
As of the DLQ PR, the Fraud Handler is expecting the second return of fraud_service.PredictFraud to be the transactions that failed, so this would break the DLQ. I'd recommend adding your call to PredictFraud to fraud_service.go instead, since this has the correct business logic for the dead letter queue already.

donglaiduan committed:
Merge branch 'main' into feat/add-aws-fraud-detector

donglaiduan committed:
Refactor FraudService Initialization and Enhance Fraud Prediction Logic

- Updated fraud_pipeline.go to use the new FraudService constructor with fraud detector integration.
- Modified the GfFraudService to include a fraud detector and refactored the PredictFraud method to utilize it.
- Improved error handling in the PredictFraud method for better logging of transaction failures.

donglaiduan committed:
mock aws fraud

donglaiduan committed:
lint

donglaiduan committed:
remove unused

jlarsen0206 commented on cmd/lambda/fraud/fraud_pipeline.go:
```diff
@@ -24,22 +27,28 @@ func main() {
 
 	context := context.Background()
 
-	awsConfig, err := config.LoadAWSConfig(context)
+	// Load AWS configuration
+	awsConfig, err := awsconfig.LoadDefaultConfig(context)
```
I'd keep the awsConfig as is - we use the config.LoadAWSConfig method to populate some other structs with values

donglaiduan committed:
revert awsConfig

wenjielee11 commented on cmd/csv_publisher/bank_transactions_data.csv:
Nuking the transaction dataset here will mean we have to fit our new transaction model to directly reflect the new dataset, especially since the publisher depended on the csv file. Please make the necessary changes in models/transaction.go and make sure our tests work end to end

wenjielee11 commented on internal/fraud_detection/fraud_detector.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// PredictFraud implements the FraudDetector interface using AWS Fraud Detector
+func (fd *GfAWSFraudDetector) PredictFraud(ctx context.Context, txn models.Transaction) (bool, error) {
```
Likewise the way you are mapping needs to change since we're using the new dataset now. With the updated transaction model you should be able to just call marshal/unmarshal here

donglaiduan committed:
edit transaction model

donglaiduan committed:
update tests to use new transaction model

donglaiduan committed:
test

jlarsen0206 commented on cmd/csv_publisher/bank_transactions_data.csv:
Does the publisher still work after the these changes?
